corehq/messaging/util.py

```diff
@@ -41,16 +49,31 @@ def rule_initiation_key_is_set(self):
     def rule_initiation_key_minutes_remaining(self):
         return (self.client.ttl(self.rule_initiation_key) // 60) or 1
 
-    def set_initial_progress(self):
+    def set_initial_progress(self, shard_count=0):
+        # shard_count is passed when tasks are run per each shard
         self.client.set(self.current_key, 0)
         self.client.set(self.total_key, 0)
+        if shard_count:
+            self.client.set(self.shard_count_key, shard_count)
         self.client.expire(self.current_key, self.key_expiry)
         self.client.expire(self.total_key, self.key_expiry)
         self.client.delete(self.rule_cancellation_key)
         self.set_rule_initiation_key()
 
+    def mark_shard_complete(self, db_alias):
+        completed_shards = self.client.get(self.completed_shards_key) or []
+        self.client.set(self.completed_shards_key, completed_shards + [db_alias])
+        self.client.expire(self.completed_shards_key, self.key_expiry)
+
+    def all_shards_completed(self):
+        completed_count = len(self.client.get(self.completed_shards_key) or [])
+        total_count = self.client.get(self.shard_count_key) or 0
+        return completed_count == total_count
```

Comment:
Again, because of the very less number of concurrent operations, practically I thought this wouldn't be a concern.

Reply:
Also, it doesn't matter, it will just try to clear a key again and set locked_for_editing to False again.

Reply:
Probably doesn't matter so much right now, but I would at least add a comment that it could return true for more than one shard. The right way would be to incorporate it into `mark_shard_complete` (and that would not be very hard).

Reply:
Sure